File: inflearnCodingTest/자료구조/크레인인형뽑기.js
/* 내 풀이 */
function solution(board, moves){
    let answer=0;
    const stack = [];
    const SIZE = board.length;
    moves.forEach(move => {
        for(let i = 0; i < SIZE; i++) {
            const tmp = board[i][move - 1];
            if(tmp !== 0) {
                if(stack.length > 0 && stack[stack.length - 1] === tmp) {
                    stack.pop();
                    answer += 2;
                } else {
                    stack.push(tmp);
                }
                board[i][move - 1] = 0;
                break;
            }
        }
    });
    
    return answer;
}

let a=[[0,0,0,0,0],
       [0,0,1,0,3],
       [0,2,5,0,1],
       [4,2,4,4,2],
       [3,5,1,3,1]];

let b=[1, 5, 3, 5, 1, 2, 1, 4];
console.log(solution(a, b));

/* 해설 풀이 */
function solution2(board, moves){
    let answer=0;
    let stack=[];
    moves.forEach(pos => {
        for(let i=0; i<board.length; i++){
            if(board[i][pos-1]!==0){
                let tmp=board[i][pos-1];
                board[i][pos-1]=0;
                if(tmp===stack[stack.length-1]){
                    stack.pop();
                    answer+=2;
                }
                else stack.push(tmp);
                break;
            }
        }
    });
                    
    return answer;
}

let a2=[[0,0,0,0,0],
       [0,0,1,0,3],
       [0,2,5,0,1],
       [4,2,4,4,2],
       [3,5,1,3,1]];

let b2=[1, 5, 3, 5, 1, 2, 1, 4];
console.log(solution2(a2, b2));

File: programmers/src/main/java/level1/kakao/크레인인형뽑기.java
package level1.kakao;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Objects;

public class 크레인인형뽑기 {

	public int solution(int[][] board, int[] moves) {
		int size = board.length;
		CustomStack stack = new CustomStack();

		for (int move : moves) {
			move = move - 1;
			for (int i = 0; i < size; i++) {
				if (board[i][move] != 0) {
					stack.game(board[i][move]);
					board[i][move] = 0;
					break;
				}
			}
		}

		return stack.getBang();
	}

	static class CustomStack {

		Deque<Integer> stack;
		int bang = 0;

		CustomStack() {
			stack = new ArrayDeque<>();
		}

		public void game(Integer i) {
			if (Objects.equals(stack.peekLast(), i)) {
				stack.removeLast();
				bang += 2;
			} else {
				stack.addLast(i);
			}
		}

		public int getBang() {
			return bang;
		}
	}

}
